api/v1beta1/openstackdataplanenodeset_webhook.go

```diff
@@ -90,10 +102,68 @@ var _ webhook.Validator = &OpenStackDataPlaneNodeSet{}
 func (r *OpenStackDataPlaneNodeSet) ValidateCreate() error {
 	openstackdataplanenodesetlog.Info("validate create", "name", r.Name)
 
-	// TODO(user): fill in your validation logic upon object creation.
+	// Currently, this check is only valid for PreProvisioned nodes. Since we can't possibly
+	// have duplicates in Baremetal Deployments, we can exit early here for Baremetal NodeSets.
+	if !r.Spec.PreProvisioned {
+		return nil
+	}
+
+	var errors field.ErrorList
+
+	nodeSetList := &OpenStackDataPlaneNodeSetList{}
+	opts := &client.ListOptions{
+		Namespace: r.ObjectMeta.Namespace,
+	}
+
+	err := webhookClient.List(context.TODO(), nodeSetList, opts)
+	if err != nil {
+		return err
+	}
+
+	// If this is the first NodeSet being created, then there can be no duplicates
+	// we can exit early here.
+	if len(nodeSetList.Items) == 0 {
+		return nil
+	}
+
+	duplicateNodesError := r.duplicateNodeCheck(nodeSetList)
+	errors = append(errors, &duplicateNodesError)
+
+	if errors != nil {
+		return apierrors.NewInvalid(
+			schema.GroupKind{Group: "dataplane.openstack.org", Kind: "OpenStackDataPlaneNodeSet"},
+			r.Name,
+			errors)
+	}
+
 	return nil
 }
 
+// duplicateNodeCheck checks the NodeSetList for pre-existing nodes. If the user is trying to redefine an
+// existing node, we will return an error and block resource creation.
+func (r *OpenStackDataPlaneNodeSet) duplicateNodeCheck(nodeSetList *OpenStackDataPlaneNodeSetList) (errors field.Error) {
+	existingNodeNames := make([]string, 0)
+	for _, existingNode := range nodeSetList.Items {
+		for _, node := range existingNode.Spec.Nodes {
+			existingNodeNames = append(existingNodeNames, node.HostName)
+			if node.Ansible.AnsibleHost != "" {
+				existingNodeNames = append(existingNodeNames, node.Ansible.AnsibleHost)
+			}
+		}
+	}
+
+	for _, newNodeName := range r.Spec.Nodes {
+		if slices.Contains(existingNodeNames, newNodeName.HostName) || slices.Contains(existingNodeNames, newNodeName.Ansible.AnsibleHost) {
+			errors = *field.Invalid(
```

Comment:
It would be useful for the user to list all the duplication instead of showing only one of them. So I would use an `field.ErrorList` instead of a `field.Error`. E.g. https://github.com/openstack-k8s-operators/nova-operator/blob/ccdb40976d65fdbc7c401ac345433ab6d357dbd0/api/v1beta1/nova_webhook.go#L143

Reply:
Probably, but this is already going to be really inefficient if we have 100's of nodes. So my intention was to just exit early as soon as we hit a duplicate, rather than continuing to iterate over any beyond the duplicate.

Reply:
Just realised now that I'm still iterating over the rest of them anyway, and this would just end up with the last error, not the first error...

Reply:
Done in the latest iteration.